phpBB/phpbb/ucp/controller/reset_password.php

```diff
@@ -0,0 +1,436 @@
+	protected $db;
+
+	/** @var dispatcher */
+	protected $dispatcher;
+
+	/** @var helper */
+	protected $helper;
+
+	/** @var language */
+	protected $language;
+
+	/** @var log_interface */
+	protected $log;
+
+	/** @var manager */
+	protected $passwords_manager;
+
+	/** @var request_interface */
+	protected $request;
+
+	/** @var template */
+	protected $template;
+
+	/** @var user */
+	protected $user;
+
+	/** @var array phpBB DB table names */
+	protected $tables;
+
+	/** @var string phpBB root path */
+	protected $root_path;
+
+	/** @var string PHP extension */
+	protected $php_ext;
+
+	/**
+	 * Reset password controller constructor.
+	 *
+	 * @param config $config
+	 * @param driver_interface $db
+	 * @param dispatcher $dispatcher
+	 * @param helper $helper
+	 * @param language $language
+	 * @param log_interface $log
+	 * @param manager $passwords_manager
+	 * @param request_interface $request
+	 * @param template $template
+	 * @param user $user
+	 * @param array $tables
+	 * @param $root_path
+	 * @param $php_ext
+	 */
+	public function __construct(config $config, driver_interface $db, dispatcher $dispatcher, helper $helper,
+								language $language, log_interface $log, manager $passwords_manager,
+								request_interface $request, template $template, user $user, $tables, $root_path, $php_ext)
+	{
+		$this->config = $config;
+		$this->db = $db;
+		$this->dispatcher = $dispatcher;
+		$this->helper = $helper;
+		$this->language = $language;
+		$this->log = $log;
+		$this->passwords_manager = $passwords_manager;
+		$this->request = $request;
+		$this->template = $template;
+		$this->user = $user;
+		$this->tables = $tables;
+		$this->root_path = $root_path;
+		$this->php_ext = $php_ext;
+	}
+
+	/**
+	 * Init controller
+	 */
+	protected function init_controller()
+	{
+		$this->language->add_lang('ucp');
+
+		if (!$this->config['allow_password_reset'])
+		{
+			trigger_error($this->language->lang('UCP_PASSWORD_RESET_DISABLED', '<a href="mailto:' . htmlspecialchars($this->config['board_contact']) . '">', '</a>'));
+		}
+	}
+
+	/**
+	 * Remove reset token for specified user
+	 *
+	 * @param int $user_id User ID
+	 */
+	protected function remove_reset_token(int $user_id)
+	{
+		$sql_ary = [
+			'reset_token'				=> '',
+			'reset_token_expiration'	=> 0,
+		];
+
+		$sql = 'UPDATE ' . $this->tables['users'] . '
+					SET ' . $this->db->sql_build_array('UPDATE', $sql_ary) . '
+					WHERE user_id = ' . $user_id;
+		$this->db->sql_query($sql);
+	}
+
+	/**
+	 * Handle password reset request
+	 *
+	 * @return Response
+	 */
+	public function request()
+	{
+		$this->init_controller();
+
+		$submit		= $this->request->is_set_post('submit');
+		$username	= $this->request->variable('username', '', true);
+		$email		= strtolower($this->request->variable('email', ''));
+
+		add_form_key('ucp_reset_password');
+
+		if ($submit)
+		{
+			if (!check_form_key('ucp_reset_password'))
+			{
+				return $this->helper->message('FORM_INVALID');
```

Comment:
You're using `$helper->message()` to show an error message.
While I understand that you are in the UCP and can not throw exceptions, I still think `trigger_error()` is currently better. Will be easier to find and replace later on with proper exceptions

Reply:
I can actually already use exceptions for this as the pages are handled via app.php